subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java

```diff
@@ -103,21 +105,50 @@ protected DefaultDomainObjectCollection(DefaultDomainObjectCollection<? super T>
     }
 
     public Iterator<T> iterator() {
+        if (constantTimeIsEmpty()) {
+            return Iterators.emptyIterator();
+        }
         return new IteratorImpl(getStore().iterator());
     }
 
     public void all(Action<? super T> action) {
+
         action = whenObjectAdded(action);
 
+        if (constantTimeIsEmpty()) {
+            return;
+        }
+
         // copy in case any actions mutate the store
         // linked list because the underlying store may preserve order
-        Collection<T> copied = new LinkedList<T>(this);
-
-        for (T t : copied) {
-            action.execute(t);
+        // We make best effort not to create an intermediate collection if this container
+        // is empty.
+        Collection<T> copied = null;
+        for (T t : this) {
+            if (copied == null) {
+                copied = Lists.newArrayListWithExpectedSize(estimatedSize());
+            }
+            copied.add(t);
+        }
+        if (copied != null) {
+            for (T t : copied) {
+                action.execute(t);
+            }
         }
     }
 
+    /**
+     * Returns true if, and only if, the store is empty AND we know that we
+     * can query its size in constant time. Otherwise, it returns false, which
+     * doesn't mean that the store is empty, but just that we cannot determine
+     * in constant time.
+     *
+     * @return true if and only if the store is empty and can tell in constant time
+     */
+    private boolean constantTimeIsEmpty() {
```

Comment:
-1 This is optimization, and it should be very clear it's an optimization. I want the method name to be clear enough to say "this is capable of returning the size in constant time". `mayHaveElements` is not telling this, and it would encourage implementors to return an answer to the question, which may not be in constant time.